### DISCUSSION TITLE
System.IO.FileNotFoundException: Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies.

### DISCUSSION THREAD

**Discussion by u-netanel** [cefsharp/CefSharp#3436] [General]
I think the issue of runtime System.IO.FileNotFoundException: 'Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.' (see details here (https://stackoverflow.com/questions/66359832/how-can-i-fix-cefsharp-problem-after-update)) is related to differing Product version numbers on CefSharp.dll and 2 other files (version 88.2.90.0) and CefSharp.Core.Runtime.dll (version 88.2.90). It may be (also) related to the version numbers appearing on the NuGet Package manager - it is 88.2.9 for Cef.redist.x86 and 88.2.90 for CefSharp.Common and CefSharp.WinForms.
In any case I am eagerly expecting the issue to be fixed.
Thank you

_Originally posted by @u-netanel in https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-793933138_

**Comment by amaitland** [comment]
As per https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-794355029

The difference in version number has nothing to do with any runtime error you are seeing. 

Are you upgrading from a previous version? 

Also please answer the questions listed https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support#when-asking-a-question-please-include-the-following

**Reply by u-netanel** [reply]
I did not explicitly upgrade from a pervious version but i did try different versions in order to solve the problem.
As mentioned i took version 88.2.90
I am using WinForms
I am targeting .Net framework 4.5.2

I finally solved the issue by running the program from a folder containing both my executable and all related dlls and resources, including CefSharp of course.

**Reply by amaitland** [reply]
If you wish to move the dependencies/resources to a subfolder then extra steps are required. You can see https://github.com/cefsharp/CefSharp/discussions/3371 for a discussion on the topic.

**Reply by Cla222** [reply]
@u-netanel , @amaitland 
I have read about your solution but I cannot understand how to implement it .... how did you do it?

the error occurs at installation, and the files (all of them, even the one that gives me problems) are all present in the "application files" folder
among other things, the error also presents it on the same pc, where CEF is correctly installed as it works in debug

thanks for your help :)

Claudio

**Reply by amaitland** [reply]
> the error occurs at installation, and the files (all of them, even the one that gives me problems) are all present in the "application files" folder
among other things, the error also presents it on the same pc, where CEF is correctly installed as it works in debug

@Cla222 If the application runs from within Visual Studio then it's likely a problem with your installer. Try copying the files from your debug folder to your install path as a test.

Please provide additional information, what installer you are using, .net version, version of CefSharp, etc.

**Reply by Cla222** [reply]
Hi @amaitland and thanks for your reply.. 

I tried to copy the files manually (green block, the red block is related to the files that were already present following the publication) but it continues to give problems .... among other things I have created a new project on purpose with only CEFSHARP , so as to circumscribe the problem, but the problem persists

OS Win10 (64bit)

the version of CEF is the latest available, I remem95, I updated it the other day)


I publish directly from VS, and I have verified that I have installed the Visual C++ 2015-2019 Redistributable version

VS 2019


Another thing...the problem is not only to install the program on another host, but also in the same where the debug is ok!

![err_inst_cef](https://user-images.githubusercontent.com/88085093/141135591-9663380c-6265-456c-a30b-a47a8ab4fc04.jpg)

**Reply by amaitland** [reply]
> I tried to copy the files manually (green block

Did you copy the folders from your packages directory? You need to copy the files from your bin folder. 

> I publish directly from VS,

If you are using Clickonce then you need to follow the instructions at https://github.com/cefsharp/CefSharp/discussions/3613#discussioncomment-842336

> Another thing...the problem is not only to install the program on another host, but also in the same where the debug is ok!

That's expected if your installer isn't copying all the required files.

**Comment by Cla222** [comment]
@amaitland

I also tried this way (see image), and for scruple also putting all the files in the main dir, where the setup.exe file is present, but it keeps giving the same error ... what am I wrong again?

I have not yet tried the modifications you suggest regarding the Click Once, I was thinking first of solving the problem in the "rough and rustic" way, which might be enough for me ... however, while I await your response on my attempt, I give it a look

as always thanks ....


![err_inst_cef_daBIN](https://user-images.githubusercontent.com/88085093/141269019-e35712e5-57ec-4ac6-9410-5c7bca85e8df.jpg)

**Reply by amaitland** [reply]
@Cla222 I'd suggest following the instructions, there are step by step details listed https://github.com/cefsharp/CefSharp/discussions/3613#discussioncomment-913061

Screenshots of your folder are largely meaningless to me as I cannot verify all files are present. See https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29 for a list

**Reply by Cla222** [reply]
> @Cla222 I'd suggest following the instructions, there are step by step details listed [#3613 (reply in thread)](https://github.com/cefsharp/CefSharp/discussions/3613#discussioncomment-913061)

Many many many thanks @amaitland !!! 
How "simple" can a solution be when you have such good help! It's ok, problem solved!

**Reply by VAVathana** [reply]
@Cla222  I'd go to the direct link, but I don't know how to solve it.
Thanks!

**Reply by Cla222** [reply]
> @Cla222 I'd go to the direct link, but I don't know how to solve it. Thanks!

@VAVathana I followed the suggestion (and the link of @amaitland ), specifically I followed this part

"Build
Close Solution
Edit CefSharp.MinimalExample.WinForms.csproj adding <CefSharpBuildAction>Content</CefSharpBuildAction> to the first PropertyGroup (Example at CefSharp.MinimalExample.WinForms.csproj#L31)
Open CefSharp.MinimalExample.sln in VS2019
Rebuild"


For me this operation has solved the problem because everything you need is collected in the publication

**Reply by VAVathana** [reply]
I'm a bit of English please apology. 
I debug my project in visual studio 2019 it is worked fine, but when I set it up as exe file and install it on my machine when I open I get an error like you. I'm not clear about your answer yet. Thanks!

**Reply by Cla222** [reply]
> I'm a bit of English please apology. I debug my project in visual studio 2019 it is worked fine, but when I set it up as exe file and install it on my machine when I open I get an error like you. I'm not clear about your answer yet. Thanks!

yes, because he doesn't drag along all the files he needs when you publish, and it's not enough to copy them manually ...

If you follow those 4-5 operations, that is....

do "Build" (from property window, when you can publish the project) and then close the Solution

Edit your .csproj (or .vbproj) - I've done this with notepad - and add the string <CefSharpBuildAction>Content</CefSharpBuildAction> 
to the first PropertyGroup (Example at CefSharp.MinimalExample.WinForms.csproj#L31)


RiOpen your project in VS2019
Rebuild

then PUBLISH



...you can solve.For me it's ok

**Reply by VAVathana** [reply]
Here is My first PropertyGroup in my(.csproj) file

 ```
 <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{E082BF18-C67A0-49CA-9C10-6AAB8AF1EE50}</ProjectGuid>
    <OutputType>WinExe</OutputType>
    <RootNamespace>webviewer</RootNamespace>
    <AssemblyName>webviewer</AssemblyName>
    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <Deterministic>true</Deterministic>
    <NuGetPackageImportStamp>
    </NuGetPackageImportStamp>
    <TargetFrameworkProfile />
  </PropertyGroup> 
```
I don't know how to add a string to it.
can you please give me an example in this case again? thanks!

**Reply by Cla222** [reply]
I think it's the same, 
the important thing is that it is between "PropertyGroup" and "/PropertyGroup"

But in the example is before closing, so for example, in your case, put it before "/PropertyGroup", AND AFTER TargetFrameworkProfile

**Reply by amaitland** [reply]
See https://github.com/cefsharp/CefSharp.MinimalExample/blob/cefsharp/95/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj#L31 for an example. 

Make sure the target machine has Visual Studio 2019 installed

**Reply by VAVathana** [reply]
I'm sorry for the skip reply about 5 days ago. 
first of all when I add <`CefSharpBuildAction>Content</CefSharpBuildAction>` it still get same error as before. 
and after that I add `<CefSharpAnyCpuSupport>true</CefSharpAnyCpuSupport>` I god an error when I debug in vs 2019 is 
**System.IO.FileNotFoundException: 'Could not load file or assembly 'CefSharp.Core.Runtime, Version=95.7.141.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies. The system cannot find the file specified.'** 

 Thanks! and I'm waiting for help.

**Reply by amaitland** [reply]
> I god an error when I debug in vs 2019

@VAVathana https://github.com/cefsharp/CefSharp.MinimalExample provides a working example. I'd suggest you test with that first. If you are still having problems then I'd suggest creating your [own discussion](https://github.com/cefsharp/CefSharp/discussions/new) and answering the questions listed at https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support#when-asking-a-question-please-include-the-following

I need more information before I can speculate what's going on.

**Reply by Ysera86** [reply]
Hello @amaitland ,
Thanks for all the support and awesome project.
When I upgrade from 88.2.9 to 106 latest release ,   -The CefSharp.Core.Runtime dll existed on 88.2.9 -  it's missing from bin folder and I get the following error. Would there be any comments you would point me to?

"Could not load file or assembly 'CefSharp.Core.Runtime, Version=106.0.290.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies."

ps: I have a wpf library project that integrates to an executable which is not mine. And on the server/client that executes the application .net framework 4.5.2  multi targeting pack and c++ 2015-2022 redist x86 and x64 are downloaded.

These are the packages I use: 
![cef packages](https://user-images.githubusercontent.com/17846150/196943245-0f70849a-0f77-46eb-afbb-b765f974f9eb.PNG)

**Reply by amaitland** [reply]
### Troubleshooting 

Add the following to your `csproj/vbproj` file at the end

```xml
<Target Name="CefSharpAfterBuildDebug" AfterTargets="AfterBuild">
    <CallTarget Targets="CefSharpAfterBuildDiagnostic" />
</Target>
```

Change your `MSBuild project output verbosity` to `Normal`:
```
1. Go to Tools->Options menu in VS
2. Open Projects and Solutions->Build and Run
3. Change the value of the MSBuild project build output verbosity. Pick Normal or higher
```

After you build, check the output window for a block of text that looks like:
```
>Target CefSharpAfterBuildDebug:
>  Target CefSharpAfterBuildDiagnostic:
>    CefSharp After Build Diagnostic
>    CefSharpBuildAction = Content
>    CefSharpPropertiesLoaded = true
>    CefSharpTargetDir =
>    CefSharpTargetDirAnyCpu32 = x86\
>    CefSharpTargetDirAnyCpu64 = x64\
>    RuntimeIdentifier = win7-x64
>    EffectivePlatform =
>    Platform = x64
>    PlatformName = x64
>    Platforms = x86;x64;AnyCPU
>    PlatformTarget = x64
>    CefSharpPlatformTarget = x64
>    PlatformTargetAsMSBuildArchitecture = x64
>    TargetFramework = net472
>    TargetFrameworkVersion = v4.7.2
>    libcef.dll exists = bin.net472\x64\Debug\net472\libcef.dll
>    CefSharpTransformXmlDllPath = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\Web\Microsoft.Web.Publishing.Tasks.dll
>    NuGetProjectStyle = PackageReference
>    NuGetToolVersion = 5.11.1
>    OutputType = WinExe
>    OutDir = bin.net472\x64\Debug\net472\
>    Prefer32Bit = false
```

Copy and paste the output here (no screenshots)

**Reply by Ysera86** [reply]
Cheers @amaitland , thanks for responding.
In my case, the block starts and ends as the following : 

![image](https://user-images.githubusercontent.com/17846150/197471681-1f5d0232-8d6d-465d-b35e-2553d770ba58.png)
1>CefSharpAfterBuildDiagnostic:
1>  CefSharp After Build Diagnostic
1>  CefSharpBuildAction = None
1>  CefSharpPropertiesLoaded = true
1>  CefSharpTargetDir = 
1>  CefSharpTargetDirAnyCpu32 = x86\
1>  CefSharpTargetDirAnyCpu64 = x64\
1>  CefSharpAnyCpuSupport = 
1>  RuntimeIdentifier = 
1>  EffectivePlatform = 
1>  Platform = AnyCPU
1>  PlatformName = AnyCPU
1>  Platforms = 
1>  PlatformTarget = 
1>  CefSharpPlatformTarget = AnyCPU
1>  PlatformTargetAsMSBuildArchitecture = CurrentArchitecture
1>  TargetFramework = 
1>  TargetFrameworkVersion = v4.7.2
1>  libcef.dll exists AnyCPU(x86) = bin\x86\libcef.dll
1>  libcef.dll exists AnyCPU(x64) = bin\x64\libcef.dll
1>  CefSharpTransformXmlDllPath = C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll
1>  NuGetProjectStyle = 
1>  NuGetToolVersion = 
1>  OutputType = Library
1>  OutDir = bin\
1>  Prefer32Bit = false

If I change to the following setting : 
![image](https://user-images.githubusercontent.com/17846150/197471575-9591361f-5229-4ecf-bf56-a3d01563b09c.png)
1>CefSharpAfterBuildDiagnostic:
1>  CefSharp After Build Diagnostic
1>  CefSharpBuildAction = None
1>  CefSharpPropertiesLoaded = true
1>  CefSharpTargetDir = 
1>  CefSharpTargetDirAnyCpu32 = x86\
1>  CefSharpTargetDirAnyCpu64 = x64\
1>  CefSharpAnyCpuSupport = 
1>  RuntimeIdentifier = 
1>  EffectivePlatform = 
1>  Platform = x86
1>  PlatformName = x86
1>  Platforms = 
1>  PlatformTarget = x86
1>  CefSharpPlatformTarget = x86
1>  PlatformTargetAsMSBuildArchitecture = x86
1>  TargetFramework = 
1>  TargetFrameworkVersion = v4.7.2
1>  libcef.dll exists = bin\x86\x86\libcef.dll
1>  CefSharpTransformXmlDllPath = C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll
1>  NuGetProjectStyle = 
1>  NuGetToolVersion = 
1>  OutputType = Library
1>  OutDir = bin\x86\x86\
1>  Prefer32Bit = false
1>
1>Build succeeded.

**Reply by amaitland** [reply]
> executable which is not mine.

What architecture does the host application target? x86, x64 or AnyCpu? 

When you upgraded the CefSharp Nuget package were any other changes made? Use your source control to check. 

Whoever setup your solution did something odd in adding x86 to the Configuration dropdown(first dropdown), usually this would be just Debug or Release.

**Reply by Ysera86** [reply]
Hey there @amaitland ,
1 - The  app is 32bit arch. on 64 bit clients.
2 -  browser2.JavascriptObjectRepository.Settings.LegacyBindingEnabled = true;   is the only change I've made, the cef upgrade needded this one.
3 - We've been using Any CPU, Debug. I've added the other setup for display purposes to your question.

**Reply by amaitland** [reply]
If your host application is x86 then I'd be building the library as x86 or AnyCpu with Prefer32Bit true.

**Reply by Ysera86** [reply]
Thank you @amaitland ! I'Ve changed AnyCpu to x86 and here it is. How could I not think of this. Anyways thanks for your time and help. Really appreciated!

**Comment by mcpbcs** [comment]
Hello, I use this instructions to compile en AnyCpu, I am using 114.2.120 version of nuget package

1.- Install X64 version of  Microsoft Visual C++ Redistributable latest supported downloads
from here (Visual Studio 2015, 2017, 2019, and 2022)
https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170


I followed this instruccions:

https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj

2.- Edit your project.csproj file and add the tags on the firts  `<PropertyGroup>`

```
<CefSharpAnyCpuSupport>true</CefSharpAnyCpuSupport>
<CefSharpBuildAction>Content</CefSharpBuildAction>
```

and (see https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj)

```
  <Target Name="CefSharpAfterBuildDebug" AfterTargets="AfterBuild">
    <CallTarget Targets="CefSharpAfterBuildDiagnostic" />
  </Target>
  <Import Project="..\packages\CefSharp.Common.114.2.120\build\CefSharp.Common.targets" Condition="Exists('..\packages\CefSharp.Common.114.2.120\build\CefSharp.Common.targets')" />
```

Added this lines to the program.cs

```
            CefRuntime.SubscribeAnyCpuAssemblyResolver();

                var settings = new CefSettings()
                {
                    //By default CefSharp will use an in-memory cache, you need to specify a Cache Folder to persist data
                    CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache")
                };
                settings.CefCommandLineArgs.Add("enable-media-stream");
                //https://peter.sh/experiments/chromium-command-line-switches/#use-fake-ui-for-media-stream
                settings.CefCommandLineArgs.Add("use-fake-ui-for-media-stream");
                //For screen sharing add (see https://bitbucket.org/chromiumembedded/cef/issues/2582/allow-run-time-handling-of-media-access#comment-58677180)
                settings.CefCommandLineArgs.Add("enable-usermedia-screen-capturing");

                //Perform dependency check to make sure all relevant resources are in our output directory.
                Cef.Initialize(settings, performDependencyCheck: true, browserProcessHandler: null);
```

!! And its working now on any cpu !!!

before close your program you need to call Shutdown.. In windows forms I use:

```
 protected override void OnClosing(CancelEventArgs e)
        {
            Cef.Shutdown();

            base.OnClosing(e);
        }
```

And.. the log:
```

17>  CefSharp After Build Diagnostic
17>  CefSharpBuildAction = Content
17>  CefSharpPropertiesLoaded = true
17>  CefSharpTargetDir = 
17>  CefSharpTargetDirAnyCpu32 = x86\
17>  CefSharpTargetDirAnyCpu64 = x64\
17>  CefSharpAnyCpuSupport = true
17>  RuntimeIdentifier = 
17>  EffectivePlatform = 
17>  Platform = AnyCPU
17>  PlatformName = AnyCPU
17>  Platforms = 
17>  PlatformTarget = AnyCPU
17>  CefSharpPlatformTarget = AnyCPU
17>  PlatformTargetAsMSBuildArchitecture = CurrentArchitecture
17>  TargetFramework = 
17>  TargetFrameworkVersion = v4.5.2
17>  libcef.dll exists AnyCPU(x86) = bin\Debug\x86\libcef.dll
17>  libcef.dll exists AnyCPU(x64) = bin\Debug\x64\libcef.dll
17>  CefSharpTransformXmlDllPath = C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll
17>  NuGetProjectStyle = 
17>  NuGetToolVersion = 
17>  OutputType = WinExe
17>  OutDir = bin\Debug\
17>  Prefer32Bit = false
```

**Reply by It-Techus2011** [reply]
Thanks for your help. 
I followed all the step as mentioned above and still getting error "Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.". I'm running on Windows 10 pro. Please let me know.

**Reply by It-Techus2011** [reply]
![image](https://github.com/cefsharp/CefSharp/assets/160270033/0daf5721-dc2f-43c4-a2bc-c5f4bb8e948a)

**Comment by BabakSouth** [comment]
Please just install this file in your machine. your issue will be resolved.
[https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170)

**Comment by It-Techus2011** [comment]
I followed all the step as mentioned above BY [mcpbcs] and still getting error "Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.". I'm running on Windows 10 pro. Please let me know.

**Reply by BabakSouth** [reply]
I don't have any problem for its dependencies. I am sure you should install vc-redist x64 on your machine . this your mentioned error message will be gone.

**Reply by It-Techus2011** [reply]
I did installed from : https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170
BUT still getting error on windows 10 pro and windows 11.

I used following Code: 
                //// Initialize CefSharp
                var settings = new CefSettings();
                Cef.Initialize(settings);

Please let me know anything wrong.
Thanks for your time.

**Comment by gustavo-cantero** [comment]
I've the same problem, were you able to solve it?

**Reply by It-Techus2011** [reply]
I installed 32 bits, which fixed it. because my .net app was compiled in x86.

X86	https://aka.ms/vs/17/release/vc_redist.x86.exe	Permalink for latest supported x86 version

**Comment by cebik** [comment]
@It-Techus2011 was right! Installing the 32-bit version of the C++ libraries solves the problem.
https://aka.ms/vs/17/release/vc_redist.x86.exe 
Permalink for latest supported x86 version

**Comment by SimonKravis** [comment]
Same problem - works in Visual Studio 2022 compiled to 64 bit, doesn't work after installation on different 64 bit Win 11. Installing 32-bit c++ library vc_redist.x86.exe on didn't fix it. Cannot build as Any CPU 32-bit preferred  as suggested as other components fail. Build for target Any CPU without 32-bit preferred also fails.

When I run Dependencies(x64) on CefSharp.core.runtime.dll on the Win 11 machine where the app fails, no dependencies show as missing even though the error 'could not load "CefSharp.core.runtime.dll" or one of its dependencies is missing' is thrown when I try to load CefSharp web browser. libcef.dll shows chrome_elf.dll as missing, but when it is added, another instance of the app is started when the CefSharp web browser is loaded on the Win 11 machine. On the Dev machine it still works OK.

**Reply by amaitland** [reply]
> compiled to 64 bit, doesn't work after installation on different 64 bit Win 11. Installing 32-bit c++ library vc_redist.x86.exe on didn't fix it.

If you compiled your app as `x64` then you need the `x64` build of `VC++ Redist`. The `Redist` must match the process architecture.
